File: config_manager.py
from PyQt6.QtCore import QSettings
from cryptography.fernet import Fernet
import base64
import os


class ConfigManager:
    def __init__(self):
        self.settings = QSettings("AutoCommit", "GitSync")
        self._init_encryption()

    def _init_encryption(self):
        """初始化加密密钥"""
        key = self.settings.value("encryption_key")
        if not key:
            key = Fernet.generate_key()
            self.settings.setValue("encryption_key", key)
        self.cipher = Fernet(key)

    def _encrypt(self, text: str) -> str:
        """加密文本"""
        if not text:
            return ""
        encrypted = self.cipher.encrypt(text.encode())
        return base64.b64encode(encrypted).decode()

    def _decrypt(self, encrypted_text: str) -> str:
        """解密文本"""
        if not encrypted_text:
            return ""
        try:
            encrypted = base64.b64decode(encrypted_text.encode())
            decrypted = self.cipher.decrypt(encrypted)
            return decrypted.decode()
        except Exception:
            return ""

    @property
    def repo_url(self) -> str:
        return self.settings.value("repo_url", "", str)

    @repo_url.setter
    def repo_url(self, value: str):
        self.settings.setValue("repo_url", value)

    @property
    def username(self) -> str:
        return self.settings.value("username", "", str)

    @username.setter
    def username(self, value: str):
        self.settings.setValue("username", value)

    @property
    def password(self) -> str:
        encrypted = self.settings.value("password", "", str)
        return self._decrypt(encrypted)

    @password.setter
    def password(self, value: str):
        self.settings.setValue("password", self._encrypt(value))

    @property
    def target_branch(self) -> str:
        return self.settings.value("target_branch", "my-sync-branch", str)

    @target_branch.setter
    def target_branch(self, value: str):
        self.settings.setValue("target_branch", value)

    @property
    def source_branch(self) -> str:
        return self.settings.value("source_branch", "main", str)

    @source_branch.setter
    def source_branch(self, value: str):
        self.settings.setValue("source_branch", value)

    @property
    def user_name(self) -> str:
        return self.settings.value("user_name", "", str)

    @user_name.setter
    def user_name(self, value: str):
        self.settings.setValue("user_name", value)

    @property
    def user_email(self) -> str:
        return self.settings.value("user_email", "", str)

    @user_email.setter
    def user_email(self, value: str):
        self.settings.setValue("user_email", value)

    @property
    def sync_time(self) -> str:
        return self.settings.value("sync_time", "00:00", str)

    @sync_time.setter
    def sync_time(self, value: str):
        self.settings.setValue("sync_time", value)

    @property
    def last_sync_hash(self) -> str:
        return self.settings.value("last_sync_hash", "", str)

    @last_sync_hash.setter
    def last_sync_hash(self, value: str):
        self.settings.setValue("last_sync_hash", value)

    @property
    def last_sync_time(self) -> str:
        return self.settings.value("last_sync_time", "从未同步", str)

    @last_sync_time.setter
    def last_sync_time(self, value: str):
        self.settings.setValue("last_sync_time", value)

    @property
    def is_configured(self) -> bool:
        """检查配置是否完整"""
        return all([
            self.repo_url,
            self.username,
            self.password,
            self.target_branch,
            self.user_name,
            self.user_email
        ])

    def sync(self):
        """同步配置到磁盘"""
        self.settings.sync()

    def clear(self):
        """清空所有配置"""
        self.settings.clear()
        self._init_encryption()

File: src/services/config-manager.js
const Store = require('electron-store');
const crypto = require('crypto');

const ALGORITHM = 'aes-256-gcm';
const KEY_LENGTH = 32;
const IV_LENGTH = 12;
const AUTH_TAG_LENGTH = 16;

class ConfigManager {
  constructor() {
    this.store = new Store({ name: 'git-sync-config' });
    this._ensureEncryptionKey();
  }

  _ensureEncryptionKey() {
    if (!this.store.get('_encryptionKey')) {
      const key = crypto.randomBytes(KEY_LENGTH).toString('hex');
      this.store.set('_encryptionKey', key);
    }
  }

  _getKey() {
    return Buffer.from(this.store.get('_encryptionKey'), 'hex');
  }

  _encrypt(plaintext) {
    if (!plaintext) return '';
    const iv = crypto.randomBytes(IV_LENGTH);
    const cipher = crypto.createCipheriv(ALGORITHM, this._getKey(), iv);
    let encrypted = cipher.update(plaintext, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    const authTag = cipher.getAuthTag().toString('hex');
    return `${iv.toString('hex')}:${authTag}:${encrypted}`;
  }

  _decrypt(ciphertext) {
    if (!ciphertext) return '';
    try {
      const [ivHex, authTagHex, encrypted] = ciphertext.split(':');
      const iv = Buffer.from(ivHex, 'hex');
      const authTag = Buffer.from(authTagHex, 'hex');
      const decipher = crypto.createDecipheriv(ALGORITHM, this._getKey(), iv);
      decipher.setAuthTag(authTag);
      let decrypted = decipher.update(encrypted, 'hex', 'utf8');
      decrypted += decipher.final('utf8');
      return decrypted;
    } catch {
      return '';
    }
  }

  get repoUrl() { return this.store.get('repoUrl', ''); }
  set repoUrl(v) { this.store.set('repoUrl', v); }

  get username() { return this.store.get('username', ''); }
  set username(v) { this.store.set('username', v); }

  get password() { return this._decrypt(this.store.get('encryptedPassword', '')); }
  set password(v) { this.store.set('encryptedPassword', this._encrypt(v)); }

  get sourceBranch() { return this.store.get('sourceBranch', 'main'); }
  set sourceBranch(v) { this.store.set('sourceBranch', v); }

  get targetBranch() { return this.store.get('targetBranch', ''); }
  set targetBranch(v) { this.store.set('targetBranch', v); }

  get userName() { return this.store.get('userName', ''); }
  set userName(v) { this.store.set('userName', v); }

  get userEmail() { return this.store.get('userEmail', ''); }
  set userEmail(v) { this.store.set('userEmail', v); }

  get syncTime() { return this.store.get('syncTime', '00:00'); }
  set syncTime(v) { this.store.set('syncTime', v); }

  get lastSyncHash() { return this.store.get('lastSyncHash', ''); }
  set lastSyncHash(v) { this.store.set('lastSyncHash', v); }

  get lastSyncTime() { return this.store.get('lastSyncTime', '从未同步'); }
  set lastSyncTime(v) { this.store.set('lastSyncTime', v); }

  get isConfigured() {
    return !!(this.repoUrl && this.username && this.password &&
              this.targetBranch && this.userName && this.userEmail);
  }

  getAll() {
    return {
      repoUrl: this.repoUrl,
      username: this.username,
      password: this.password,
      sourceBranch: this.sourceBranch,
      targetBranch: this.targetBranch,
      userName: this.userName,
      userEmail: this.userEmail,
      syncTime: this.syncTime,
      lastSyncHash: this.lastSyncHash,
      lastSyncTime: this.lastSyncTime,
      isConfigured: this.isConfigured,
    };
  }

  saveAll(data) {
    if (data.repoUrl !== undefined) this.repoUrl = data.repoUrl;
    if (data.username !== undefined) this.username = data.username;
    if (data.password !== undefined) this.password = data.password;
    if (data.sourceBranch !== undefined) this.sourceBranch = data.sourceBranch;
    if (data.targetBranch !== undefined) this.targetBranch = data.targetBranch;
    if (data.userName !== undefined) this.userName = data.userName;
    if (data.userEmail !== undefined) this.userEmail = data.userEmail;
    if (data.syncTime !== undefined) this.syncTime = data.syncTime;
  }
}

module.exports = { ConfigManager };
